internal/webhook/template_webhook_test.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2024
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package webhook
+
+import (
+	"context"
+	"testing"
+
+	. "github.com/onsi/gomega"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	"github.com/Mirantis/hmc/api/v1alpha1"
+	"github.com/Mirantis/hmc/internal/controller"
+	"github.com/Mirantis/hmc/test/objects/managedcluster"
+	"github.com/Mirantis/hmc/test/objects/template"
+	"github.com/Mirantis/hmc/test/scheme"
+)
+
+func TestClusterTemplateValidateDelete(t *testing.T) {
+	g := NewWithT(t)
+	ctx := context.Background()
+	tpl := template.NewClusterTemplate(template.WithName("testTemplateFail"))
+
+	tests := []struct {
+		name            string
+		template        *v1alpha1.ClusterTemplate
+		existingObjects []runtime.Object
+		err             string
+		warnings        admission.Warnings
+	}{
+		{
+			name:            "should fail if ManagedCluster objects exist",
+			template:        tpl,
+			existingObjects: []runtime.Object{managedcluster.NewManagedCluster(managedcluster.WithTemplate(tpl.Name))},
+			warnings:        admission.Warnings{"The ClusterTemplate object can't be removed if ManagedCluster objects still exist"},
+			err:             "template deletion is forbidden",
+		},
+		{
+			name:            "should be OK because of a different cluster",
+			template:        tpl,
+			existingObjects: []runtime.Object{managedcluster.NewManagedCluster()},
+		},
+		{
+			name:            "should succeed",
+			template:        template.NewClusterTemplate(),
+			existingObjects: []runtime.Object{managedcluster.NewManagedCluster()},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			c := fake.NewClientBuilder().
+				WithScheme(scheme.Scheme).
+				WithRuntimeObjects(tt.existingObjects...).
+				WithIndex(tt.existingObjects[0], v1alpha1.TemplateKey, controller.ExtractTemplateName).
```

Comment:
If we pass a wrong object then line 519 of [internal/controller/managedcluster_controller.go](https://github.com/Mirantis/hmc/pull/301/files#diff-f79d441347dc2389fb8d9662097d55099d962aee056e150a45bcf767a9d3445c) will handle this. It will be a test for that part of code.